hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -321,6 +324,11 @@ def test_function(self, data):
                     changed = True
 
             if changed:
+                if key not in self.interesting_examples:
+                    self.examples_at_last_bug = self.valid_examples
+                    self.finish_if_after_time = min([
+                        self.finish_if_after_time, benchmark_time() + 10
+                    ])
                 self.save_buffer(data.buffer)
```

Comment:
I'm not a fan of the late-termination policy being spread across both `test_function` and `generate_new_examples`.

It might be better to do one of the following:

- Have `test_function` record a timestamp and example count, and then make `generate_new_examples` enforce the specific policies of "10 minutes" and "twice as many examples".

- Have `test_function` use the policies to compute a deadline and an example limit, and then have `generate_new_examples` just check whether those have been exceeded.

Or encapsulate all of the logic in a dedicated helper class, and call into it from both code sites.